Mathlib/Analysis/Complex/Residue/Basic.lean

```diff
@@ -0,0 +1,383 @@
+/-
+Copyright (c) 2025 Roman Kvasnytskyi. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Roman Kvasnytskyi
+-/
+import Mathlib.Analysis.Complex.CauchyIntegral
+
+/-!
+# Residues at isolated singularities
+
+We define the residue of a complex function at an isolated singularity and prove basic
+properties. The residue is the coefficient of `(z - c)⁻¹` in the Laurent expansion around `c`,
+which can be computed as `(2πi)⁻¹` times a circle integral.
+
+## Main definitions
+
+* `HasIsolatedSingularityAt f c`: `f` has an isolated singularity at `c`
+* `HasRemovableSingularityAt f c`: `f` has a removable singularity at `c`
+* `residue f c hf`: the residue of `f` at `c`
+
+## Main results
+
+* `residue_eq_two_pi_I_inv_smul_circleIntegral`: radius independence for residues
+* `residue_of_holomorphic`: holomorphic functions have zero residue
+* `residue_simple_pole`: residue formula for simple poles
+-/
+
+noncomputable section
+
+open scoped Real Topology
+
+open Set Metric Complex Filter
+
+namespace Complex
+
+/-- `f` has an isolated singularity at `c` if `f` is holomorphic in a punctured
+neighborhood of `c`. -/
+def HasIsolatedSingularityAt (f : ℂ → ℂ) (c : ℂ) : Prop :=
+  ∃ R > 0, DifferentiableOn ℂ f (ball c R \ {c})
+
+/-- `f` has a removable singularity at `c` if it has an isolated singularity there and
+extends to a holomorphic function in a full neighborhood of `c`. -/
+def HasRemovableSingularityAt (f : ℂ → ℂ) (c : ℂ) : Prop :=
+  HasIsolatedSingularityAt f c ∧ ∃ (g : ℂ → ℂ) (R : ℝ) (_ : 0 < R),
+    DifferentiableOn ℂ g (ball c R) ∧ EqOn f g (ball c R \ {c})
+
+namespace HasIsolatedSingularityAt
+
+variable {f g : ℂ → ℂ} {c : ℂ}
```

Comment:
Does the codomain need to be ℂ? I see that in [Complex.circleIntegral_eq_of_differentiable_on_annulus_off_countable](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Analysis/Complex/CauchyIntegral.html#Complex.circleIntegral_eq_of_differentiable_on_annulus_off_countable) for example, it is more general

Reply:
Indeed, it should be possible to work with functions `f : ℂ → E` where `E` is a normed `ℂ`-vector space (some results will need completeness of `E`).

Reply:
Agreed. I'll generalize to f : ℂ → E where E is a normed ℂ-vector space. Will add [CompleteSpace E] hypothesis where needed.